{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import base64\n",
    "import bz2\n",
    "from collections import defaultdict\n",
    "import json\n",
    "import time\n",
    "\n",
    "import pandas as pd\n",
    "import requests\n",
    "\n",
    "\n",
    "COUNT_OFFTARGETS_URL = 'https://b7bc9zf9oj.execute-api.ap-southeast-2.amazonaws.com/prod/countOfftargets'\n",
    "FIND_TARGETS_URL = 'https://b7bc9zf9oj.execute-api.ap-southeast-2.amazonaws.com/prod/findTargets'\n",
    "MAX_TARGETS_PER_SCAN = 10000\n",
    "\n",
    "\n",
    "SPECIES_S3_LOCATIONS = {\n",
    "    'B. pertussis': ('cdiag-genomes', 'BordetellaPertussis.fasta'),\n",
    "    'Ebola': ('cdiag-genomes', 'Ebola.fasta'),\n",
    "    'F. Solani': ('cdiag-genomes', 'FusariumSolani.fasta'),\n",
    "    'H1N1 1': ('cdiag-genomes', 'H1N1_1.fa'),\n",
    "    'H1N1 2': ('cdiag-genomes', 'H1N1_2.fa'),\n",
    "    'H1N1 3': ('cdiag-genomes', 'H1N1_3.fa'),\n",
    "    'H1N1 4': ('cdiag-genomes', 'H1N1_4.fa'),\n",
    "    'H1N1 5': ('cdiag-genomes', 'H1N1_5.fa'),\n",
    "    'H1N1 6': ('cdiag-genomes', 'H1N1_6.fa'),\n",
    "    'H1N1 7': ('cdiag-genomes', 'H1N1_7.fa'),\n",
    "    'H1N1 8': ('cdiag-genomes', 'H1N1_8.fa'),\n",
    "    'hMPV': ('cdiag-genomes', 'humanMetapneumovirus.fasta'),\n",
    "    'L. prolificans': ('cdiag-genomes', 'LomentosporaProlificans.fasta'),\n",
    "    'RSV': ('cdiag-genomes', 'respiratorySyncytialVirus.fasta'),\n",
    "    'S. aureus': ('cdiag-genomes', 'StaphylococcusAureus.fasta'),\n",
    "    'Coronavirus': ('cdiag-genomes', 'Coronavirus.fasta'),\n",
    "}\n",
    "\n",
    "\n",
    "def check_request(method, *args, **kwargs):\n",
    "    print(f'{method}: {args[0]}')\n",
    "    now = time.time()\n",
    "    response = requests.request(method, *args, **kwargs)\n",
    "    print(f'Completed request in {round((time.time() - now) * 1000)}ms')\n",
    "    try:\n",
    "        response.raise_for_status()\n",
    "    except requests.HTTPError as e:\n",
    "        print(f'HTTP Error {e.response.status_code}')\n",
    "        print(e.response.text)\n",
    "        raise\n",
    "    return response\n",
    "\n",
    "\n",
    "def format_data(data):\n",
    "    num_organisms = len(data[0][-1])\n",
    "    num_mismatch_columns = len(list(data[0][-1].values())[0])\n",
    "    column_names = [\"Organism\", \"Contig\", \"Target Start\", \"Strand\", \"Target\"]\n",
    "    off_target_columns = ['MM_{}'.format(i) for i in range(num_mismatch_columns)]\n",
    "    for i in range(num_organisms):\n",
    "        column_names += [\"Off-Target Organism\"]\n",
    "        column_names += off_target_columns\n",
    "    formatted_data = []\n",
    "    for element in data:\n",
    "        toAppend = element[:5]\n",
    "        for organism in element[5]:\n",
    "            toAppend.append(organism)\n",
    "            for target in element[5][organism]:\n",
    "                toAppend.append(target)\n",
    "        formatted_data.append(toAppend)\n",
    "    return pd.DataFrame.from_records(formatted_data, columns = column_names)\n",
    "\n",
    "\n",
    "def get_all_offtargets(target_positions, organisms, rule, max_mismatches):\n",
    "    target_mismatches = defaultdict(dict)\n",
    "    targets = list(target_positions)\n",
    "    for offtarget_organism in organisms:\n",
    "        bucket, key = SPECIES_S3_LOCATIONS[offtarget_organism]\n",
    "        start = 0\n",
    "        end = MAX_TARGETS_PER_SCAN\n",
    "        while True:\n",
    "            target_chunk = targets[start:end]\n",
    "            if not target_chunk:\n",
    "                break\n",
    "            target_hits_response = check_request('POST', COUNT_OFFTARGETS_URL, json={\n",
    "                'fastaBucket': bucket,\n",
    "                'fastaKey': key,\n",
    "                'mismatches': max_mismatches,\n",
    "                'rule': rule,\n",
    "                'targets': target_chunk, \n",
    "            })\n",
    "            target_hits = target_hits_response.json()\n",
    "            for i, mismatches in enumerate(target_hits):\n",
    "                seq = target_chunk[i]\n",
    "                target_mismatches[seq][offtarget_organism] = mismatches\n",
    "            start = end\n",
    "            end += MAX_TARGETS_PER_SCAN\n",
    "    output = []\n",
    "    for seq, details in target_positions.items():\n",
    "        mismatches = target_mismatches[seq]\n",
    "        for pos in details:\n",
    "            output.append(pos + [seq, mismatches])\n",
    "    return output\n",
    "\n",
    "\n",
    "def get_common_targets(in_organisms, out_organisms, rule):\n",
    "    find_targets_response = check_request('POST', FIND_TARGETS_URL, json={\n",
    "        'in_organisms': {\n",
    "            organism: {\n",
    "                's3_location': SPECIES_S3_LOCATIONS[organism],\n",
    "                'regions': regions,\n",
    "            }\n",
    "            for organism, regions in in_organisms.items()\n",
    "        },\n",
    "        'out_organisms': {\n",
    "            organism: {\n",
    "                's3_location': SPECIES_S3_LOCATIONS[organism],\n",
    "                'regions': regions,\n",
    "            }\n",
    "            for organism, regions in out_organisms.items()\n",
    "        },\n",
    "        'rule': rule,\n",
    "    })\n",
    "    json_data = bz2.decompress(base64.b64decode(find_targets_response.text))\n",
    "    common_targets = json.loads(json_data)\n",
    "    return common_targets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "IN_ORGANISMS = {\n",
    "    'B. pertussis': [\n",
    "        'CP017403.1:200000-',\n",
    "        'CP017403.1:400000-500000',\n",
    "        'CP017403.1:-100000',\n",
    "    ],\n",
    "    'S. aureus': [\n",
    "        'NC_007795.1',\n",
    "    ]\n",
    "}\n",
    "OUT_ORGANISMS = {\n",
    "    'Coronavirus': None\n",
    "}\n",
    "\n",
    "RULE = 'xxxxxxxxxxxxxxxxxxxNGG'\n",
    "MAX_MISMATCHES = 6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_targets = get_common_targets(IN_ORGANISMS, OUT_ORGANISMS, RULE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_targets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_organisms = list(IN_ORGANISMS) + list(OUT_ORGANISMS)\n",
    "start = time.time()\n",
    "offtarget_counts = get_all_offtargets(filtered_targets, all_organisms, RULE, MAX_MISMATCHES)\n",
    "print(time.time()-start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = format_data(offtarget_counts)\n",
    "df.sort_values(by=['Target'])\n",
    "with pd.option_context('display.max_rows', None, 'display.max_columns', None):\n",
    "    display(df)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
